### PR TITLE
Refactor bitbucket collector to not swallow errors & have nicer logging

### LINKED ISSUE #464: Bitbucket Exporter can hide important errors
While working through a custom CA issue with the bitbucket exporter, I saw that the first contact with the bitbucket server is in the api version check. I noticed it just swallows any exception from the request! While we get a "graceful" failure where it just returns the metric with no commit time, there's no logging done.

Even worse, it's pokémon error handling-- gotta catch 'em all, even if the errors should be more apparent. The only exceptions we should be catching are certificate verification errors and response errors (and the latter case is already handled through checking, not throwing).

### PR DESCRIPTION
**Pull request by KevinMGranger**
resolves #464

## Testing Instructions

Positive test:

1. Make a pelorus repo on bitbucket
2. Make an app password in bitbucket
3. Deploy a bitbucket committime collector using that bitbucket repo and credentials
4. Confirm that the metrics are logged / collected correctly

SSL Error test:
1. Keep the existing repo and deployment
2. Set up mockoon, enable TLS
3. Locally change the code so `git_server` is always overwritten with the local mockoon url
4. Verify the TLS/SSL error in the logs

Now, the negative test:

0. Remove the hardcoded git_server from the above test
1. Delete that repo
2. Redeploy the pod (don't rebuild!)
3. Confirm that the logs report the error and are fully detailed.

### CODE REVIEW

File: exporters/committime/collector_bitbucket.py [hunk 1]
```diff
@@ -147,22 +141,86 @@ def get_commit_time(self, metric):
...
-            # create the full URL
             url = pelorus.url_joiner(api_server, path)
-            # send a GET
+
             response = self.__session.request(
                 "GET", url=url, headers=self.DEFAULT_HEADERS
             )
             response.encoding = "utf-8"
-            api_response = response
+            response.raise_for_status()
```
Comment (KevinMGranger) [inline exporters/committime/collector_bitbucket.py:174]: The original code checked if the response was exactly 200, while this just checks if it's `400 <= status <= 599`.

I can't think of a case where this matters, but it's worth pointing out.

**Comment by KevinMGranger** [comment]
Found a strange issue when testing TLS error logging.

**Comment by mpryc** [comment]
While testing on a public bitbucket repo with no TOKEN (token not required):

Set annotations on the build, note this works fine with GIThub provider:
```
io.openshift.build.source-location=https://bitbucket.org/michalpryc/pelorus-bitbucket.git
io.openshift.build.commit.id=e2c4ef00468dfc10aad1bd2d4c9d470160a7f471
```

Exporter:

```
07-19-2022 15:40:19 DEBUG    /home/migi/Development/Konveyor/Pelorus/pelorus-minio/exporters/committime/collector_base.py:343 _set_commit_timestamp() sha: e2c4ef00468dfc10aad1bd2d4c9d470160a7f471, commit_timestamp not found in cache, executing API call.
07-19-2022 15:40:19 WARNING  Skipping non GitHub server, found bitbucket.org
07-19-2022 15:40:19 WARNING  Build first/python-binary-build1-1 in app python-binary-build1 is missing required attributes to collect data. Skipping.
07-19-2022 15:40:19 DEBUG    /home/migi/Development/Konveyor/Pelorus/pelorus-minio/exporters/committime/collector_base.py:234 get_metric_from_build() 'NoneType' object has no attribute 'commit_time'
Traceback (most recent call last):
  File "/home/migi/Development/Konveyor/Pelorus/pelorus-minio/exporters/committime/collector_base.py", line 214, in get_metric_from_build
    metric = self._set_commit_timestamp(metric, errors)
  File "/home/migi/Development/Konveyor/Pelorus/pelorus-minio/exporters/committime/collector_base.py", line 349, in _set_commit_timestamp
    if metric.commit_time is None:
```

**Comment by mpryc** [comment]
Setting GIT_PROVIDER=bitbucket works slightly better, still not able to get the commit hash timestamp. I wonder if this is expected as public repo still requires token:

```
07-19-2022 16:14:03 DEBUG    /home/migi/Development/Konveyor/Pelorus/pelorus-minio/exporters/committime/__init__.py:104 __parse_repourl() Parsed: Parsed(pathname='/michalpryc/pelorus-bitbucket', protocols=['http'], protocol='http', href='http://bitbucket.org/michalpryc/pelorus-bitbucket', resource='bitbucket.org', user=None, port=None, name='pelorus-bitbucket', owner='michalpryc')
07-19-2022 16:14:03 DEBUG    /home/migi/Development/Konveyor/Pelorus/pelorus-minio/exporters/committime/collector_base.py:288 _set_repo_url() Repo URL for build python-binary-build1-1 provided by 'io.openshift.build.source-location' annotation: http://bitbucket.org/michalpryc/pelorus-bitbucket
07-19-2022 16:14:03 DEBUG    /home/migi/Development/Konveyor/Pelorus/pelorus-minio/exporters/committime/collector_base.py:252 _set_commit_hash_from_annotations() Commit hash for build python-binary-build1-1 provided by 'io.openshift.build.commit.id' annotation: e2c4ef00468dfc10aad1bd2d4c9d470160a7f471
07-19-2022 16:14:03 DEBUG    /home/migi/Development/Konveyor/Pelorus/pelorus-minio/exporters/committime/collector_base.py:343 _set_commit_timestamp() sha: e2c4ef00468dfc10aad1bd2d4c9d470160a7f471, commit_timestamp not found in cache, executing API call.
07-19-2022 16:14:04 ERROR    While testing for API Version V_2_0 at server http://bitbucket.org, got response: 401 Client Error: Unauthorized for url: https://bitbucket.org/api/2.0/repositories
```

**Comment by mpryc** [comment]
```
07-19-2022 20:45:16 INFO     No USER and no TOKEN given. This is okay for public repositories only.
07-19-2022 20:45:16 INFO     =====Using BitBucket Collector=====

07-19-2022 20:45:16 DEBUG    /home/migi/Development/Konveyor/Pelorus/pelorus-minio/exporters/committime/collector_base.py:288 _set_repo_url() Repo URL for build python-binary-build1-1 provided by 'io.openshift.build.source-location' annotation: http://bitbucket.org/michalpryc/pelorus-bitbucket
07-19-2022 20:45:16 DEBUG    /home/migi/Development/Konveyor/Pelorus/pelorus-minio/exporters/committime/collector_base.py:252 _set_commit_hash_from_annotations() Commit hash for build python-binary-build1-1 provided by 'io.openshift.build.commit.id' annotation: e2c4ef00468dfc10aad1bd2d4c9d470160a7f471
07-19-2022 20:45:16 DEBUG    /home/migi/Development/Konveyor/Pelorus/pelorus-minio/exporters/committime/collector_base.py:343 _set_commit_timestamp() sha: e2c4ef00468dfc10aad1bd2d4c9d470160a7f471, commit_timestamp not found in cache, executing API call.
07-19-2022 20:45:17 ERROR    While testing for API Version V_2_0 at server http://bitbucket.org, got response: 401 Client Error: Unauthorized for url: https://bitbucket.org/api/2.0/repositories
```

**Comment by KevinMGranger** [comment]
I'll have to retry this on my end. May ask you for an environment variable dump.

**Comment by mpryc** [comment]
@KevinMGranger 

Here is my env, note that I was unsetting USER /USERNAME for testing this.

[env.txt](https://github.com/konveyor/pelorus/files/9148306/env.txt)

**Comment by mpryc** [comment]
@KevinMGranger the api allows to get calls to an unauthorized user based on the group / repo_name and the commit hash, all this information is available from the annotation / build url, so it should work imo. Let's discuss offline during daily sync.

Please see the api URL:

https://bitbucket.org/api/2.0/repositories/michalpryc/pelorus-bitbucket/commit/e2c4ef00468dfc10aad1bd2d4c9d470160a7f471

**Comment by KevinMGranger** [comment]
Ah, so it's just the "version check" that uses repositories that's broken.

Either we could eliminate the version check and fall back to v1 on a per-repo basis, or find an endpoint that doesn't require auth.

However, I'm willing to bet most users will use private repos.

**Comment by mpryc** [comment]
@KevinMGranger that makes me think, should we remove v1.0 api all together?

https://developer.atlassian.com/cloud/bitbucket/deprecation-notice-v1-apis/

**Comment by KevinMGranger** [comment]
I suppose it's possible that someone could be using a hosted bitbucket instance that is so old it can only do v1.0. We don't have telemetry to tell us.

However, we can do a scream test, and can always add it back later.

**Comment by KevinMGranger** [comment]
Rebased on top of #556 to simplify testing. View that as a prerequisite.

**Comment by mpryc** [comment]
Code was tested using annotations:

1. In the pelorus dir create app.py
```
cat <<EOF > app.py
#!/usr/bin/env python3
print("Hello World")
EOF

chmod +x app.py
```

2. Create namespace for build and run the build, annotate with existing bitbucket url and hash
```
export NS=buildtest
export BUILD_NAME=bitbucket-binary-build

oc create namespace "${NS}"
oc config set-context --current --namespace "${NS}"

oc new-build python --name="${BUILD_NAME}" --binary=true
oc label bc "${BUILD_NAME}" app.kubernetes.io/name="${BUILD_NAME}"

oc start-build bc/"${BUILD_NAME}" --from-file=./app.py --follow

oc annotate build "${BUILD_NAME}"-1 --overwrite \
io.openshift.build.commit.id=e2c4ef00468dfc10aad1bd2d4c9d470160a7f471 \
io.openshift.build.source-location=https://bitbucket.org/michalpryc/pelorus-bitbucket
```

3. Run the bitbucket collector with the given build/namespace
```
export PROVIDER=bitbucket
export NAMESPACES="${NS}"
python exporters/committime/app.py
```

4. Confirm the timestamp was correctly "collected"

```
INFO     Collected commit_timestamp{ namespace=buildtest, app=bitbucket-binary-build, commit=e2c4ef00468dfc10aad1bd2d4c9d470160a7f471, image_sha=sha256:cc4e6a5c159edb6477d7b07f8a82a1708201a4a218151fd1f1616d7a2daeafaa } 1657806414.0
```

**Comment by mpryc** [comment]
Similarly to previous comment tested with API_USER and TOKEN, have some comments around it.

Bitbucket requires *both* API_USER and TOKEN to be set. If API_USER or TOKEN is set and second is unset, then it will fall back to anonymous authentication without giving error:

```
401 Client Error: Unauthorized for url: https://bitbucket.org/api/2.0/repositories/michalpryc/pelorus-bitbucket/commit/e2c4ef00468dfc10aad1bd2d4c9d470160a7f471
```

I think we should check if both are set or both are unset, otherwise users will silently use non-authorized API calls even if they requested authorized ones.

**Comment by mpryc** [comment]
@KevinMGranger please let me know what do you think around my last comment, other then that everything looks perfect.

**Comment by KevinMGranger** [comment]
Very good point. Will fix that today.

**Pull request merged**